plugins/wacom-i2c/fu-wac-emr-device.c

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <gio/gio.h>
+
+#include "fu-chunk.h"
+#include "fu-wac-common.h"
+#include "fu-wac-emr-device.h"
+
+#define FU_WAC_EMR_BL_PID			0x012b
+
+struct _FuWacEmrDevice {
+	FuWacDevice		 parent_instance;
+};
+
+G_DEFINE_TYPE (FuWacEmrDevice, fu_wac_emr_device, FU_TYPE_WAC_DEVICE)
+
+static gboolean
+fu_wac_emr_device_setup (FuDevice *device, GError **error)
+{
+	FuWacEmrDevice *self = FU_WAC_EMR_DEVICE (device);
+	FuWacHidDescriptor desc;
+	guint8 data[19] = { 0 };
+	guint16 fw_ver = 0;
+	g_autofree gchar *version = NULL;
+
+	/* check PID is correct */
+	if (!fu_wac_device_get_hid_desc (FU_WAC_DEVICE (self), &desc, error))
+		return FALSE;
+	if (desc.wVendorID != FU_WAC_DEVICE_VID) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "device VID not recognised");
+		return FALSE;
+	}
+
+	/* get firmware version */
+	if (desc.wProductID == FU_WAC_EMR_BL_PID) {
+		fu_device_add_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	} else {
+		if (!fu_wac_device_get_feature (FU_WAC_DEVICE (self), 0x03,
+						data, sizeof(data), error))
+			return FALSE;
+		fw_ver = fu_common_read_uint16 (data + 11, G_LITTLE_ENDIAN);
+		fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	}
+	version = fu_common_version_from_uint16 (fw_ver, FU_VERSION_FORMAT_PAIR);
+	fu_device_set_version (device, version);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_wac_emr_device_check_mpu_type (FuWacEmrDevice *self, GError **error)
+{
+	guint8 rsp[RSP_SIZE];
+	guint8 cmd[] = {
+		FU_WAC_DEVICE_BL_REPORT_ID_CMD,
+		FU_WAC_DEVICE_BL_CMD_GET_MPUTYPE,
+		0x07,
+	};
+	if (!fu_wac_device_cmd (FU_WAC_DEVICE (self),
+				cmd, sizeof(cmd), rsp, sizeof(rsp), 0,
+				FU_WAC_DEVICE_CMD_FLAG_NONE, error)) {
+		g_prefix_error (error, "failed to get MPU type: ");
+		return FALSE;
+	}
+	if (rsp[RTRN_RSP] != 0x2e) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "MPU is not W9013 : 0x%x",
+			     rsp[RTRN_RSP]);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static guint8
```

Comment:
Shouldn't this be bigger than `guint8` so that you don't overflow?

Reply:
Nahh, it needs to overflow :/